### ISSUE TITLE
Create a static version of the IfcOpenShell hl Python API

### ISSUE THREAD

**Issue by aothms** [IfcOpenShell/IfcOpenShell#2693]
We have a wonderful high level API, kudos to everybody involved.

We were also all in agreement with the `ifcopenshell.api.run("module.function", ...args)` kind of approach. By centrally invoking all API functionality through a central function we eliminate boiler-plate within the individual API functions and we are in really good shape to run API calls e.g collaboratively on a central server!

However, I think the python ecosystem is more and more evolving towards static type decorations. Our approach makes the required arguments of a call completely opaque at the call site so we can't benefit from the static analysis built into IDEs these days. And, tbh, we've seen quite a couple of test cases failing lately due to mismatching in arguments passed to API functions.

Here's a plan. Why don't we at the cmake install step [0] generate [1] [2] static equivalents of all the API usecases?

Thoughts?

#### `api/aggregate/assign_object.py`

```python
class Usecase:
    def __init__(self, file, product=None, relating_object=None):
        ...
```

#### `api/aggregate/__init__.py`

```python
# existing content
# ...

# appended at install
def assign_object(file : ifcopenshell.file, product : ifcopenshell.entity_instance = None, relating_object : ifcopenshell.entity_instance = None) -> ifcopenshell.entity_instance:
    # [doc_string]
    ifcopenshell.api.run("aggregate.assign_object", file, product=product, relating_object=relating_object)
```

[0] https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcwrap/CMakeLists.txt#L95

[1] I think that would be `install(SCRIPT` or `install(CODE`, but maybe `add_custom_command(<target> POST_BUILD)` or `add_custom_target()` + `add_dependency()` would also work?

[2] For python parsing is it `ast.parse()` or is there sth more convenient specific-purpose we can use?

**Comment by sanzoghenzo** [comment]
Late to the party, as I totally disagree with the api/usecase approach... And [I'm not the only one](https://www.youtube.com/watch?v=o9pEzgHorH0) - A class with only an init and a single method has no reason to exist. And building it "just in case" is not useful, especially to new library users that have to figure out the magic behind `ifcopenshell.api.run`. But anyway...

It seems to me that it would be better to do the opposite: extract the api code to typed, standalone functions and call them from the `Usecase`s.
That way the `importlib`+`getattr(getattr(getattr()))` is only executed by the programs that really need it.

**Comment by aothms** [comment]
I guess the more pythonic way would have been to use decorators on free functions. The decorator could read some global state to determine to operate on a local file or online context and pass that to the api function. Is there less magic to figure out? Yes and no: people familiar to python can probably understand the implications quickly, but we have a very heterogenous crowd here with varying levels of understanding of python and programming in general. For them decorators and higher order functions will likely be more difficult to grasp and a more explicit declarative approach using classes easier to follow along.

Personally I find the incurred boiler plate on the class-based approach fairly transparent. It doesn't bother me. While I think the advise in the youtube clip is generally sound: Classes with encapsulated state are harder to reason about than pure functions. But I think that statement is more an argument on structure than syntax. For example for our express rule implementation we also use functors (classes with a __call__ method) to enrich the "function" with additional parameters that help with execution and reporting [0]. Maybe people would argue to encapsulate that information in docstrings or dictionaries or subpackages, but to me uplifting the function to a functor with member variables is the most straightforward option, but is functionally still a pure function. I see a parallel here with the hl api calls we're discussing here. Hence my argument to apply this advise to the structure of the code and not the syntax.

But as the issue indicates, we're still looking for a way to have the best of both worlds. So your input and contributions would be appreciated :)

[0] https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/express/rules/IFC2X3.py#L3553-L3560

**Comment by Moult** [comment]
Ah I should've mentioned there was one more use of the api.run(): the ability to have pre/post event listeners.

> Most of the code can be rewritten by passing the file and the settings as parameters - better yet, pass only the needed settings, so there's no need to retrieve values from a dictionary and static type checking is happier (=find bugs quickly).

Are you sure you're using the latest IfcOpenShell? We already pass the file and settings as parameters, there is no dictionary being passed as an argument anymore.

**Comment by sanzoghenzo** [comment]
> Ah I should've mentioned there was one more use of the api.run(): the ability to have pre/post event listeners.

Usually, in the observer pattern, the events are fired directly by the method/function, so it has control of the data sent to the listeners (and it also can fire events at any point of the execution); if it is not needed, a decorator can do the same thing as the current code, wrapping the function between the events.
This could be the (untested) decorator:

```python
import functools
import inspect

def signal_start_end(func):
    func_path = f"{inspect.getmodule(func).__name__}.{func.__name__}"
    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        for listener in pre_listeners.get(func_path, {}).values():
            listener(func_path, *args, *kwargs)
        func(*args, **kwargs)
        for listener in post_listeners.get(func_path, {}).values():
            listener(func_path, *args, *kwargs)
    return wrapper
```

One can use event handling libraries like pydispatch to avoid reinventing the weel... but this gets the work done.
No need to have the `should_run_listeners` argument, just don't decorate the function.

> > Most of the code can be rewritten by passing the file and the settings as parameters - better yet, pass only the needed settings, so there's no need to retrieve values from a dictionary and static type checking is happier (=find bugs quickly).
> 
> Are you sure you're using the latest IfcOpenShell? We already pass the file and settings as parameters, there is no dictionary being passed as an argument anymore.

Sorry I wasn't clear, but I was talking about the `Usecase` internals with the `self.settings` dictionary.
There's no need to use the `file` and `settings` attributes just to lazily pass the arguments to the `execute` method, since it is stateless (you talked about state, but nothing persists after an `execute` call via `api.run`).

Let me be clear: even though this architecture is useless _to me_, if you think that it brings added value (or will bring, now it is just added overhead - what's with the `setting_vcs` serialization stuff?) keep it.

My main point is that the business logic inside the `Usecase` classes should be available directly.
Then, `Usecase` can call the function to augment its functionality (even though this can be handled with decorators and FastAPI - but it's ok to keep it to avoid breaking changes)

Having the (should I say middle level api?) funcions helps with more concise unit testing, play well with static checkers and IDEs autoconpletion, and I think is a better UX for python developers.

**Comment by sanzoghenzo** [comment]
After some documenting, I can see the value of the Command Pattern you delineated.

I must confess that I did something similar (without realizing I was using that pattern) in C# for batch editing dwg files!
In my case the `BatchJob` class is initialized with only the settings, and the file is passed by the `BatchProcessor` to `BatchJob`'s `execute` method as argument.
It worked well for non-interactive tasks where you can set the objects to work on by their type, property value, and so on.

I may be missing something about your envisioned architecture, because I still fail to see how would you handle some (or many?) of the `Usecases` in a client-server context, where the client, that has to create the class instance, has no access to the `file` or its `entity_instance`s (I suppose).

BTW: many of my `execute` methods of the C# project have become lambda functions that calls some other business logic methods, so I suppose I'm biased towards lo level apis :wink:

**Comment by cvillagrasa** [comment]
Super interesting discussion guys. 

I'm not going to delve into the technical details (there's already a lot of level here), but I just wanted to say I'm totally in for making the API compatible with the static analysis from IDEs.

Besides, if we also thought of users trying ifcopenshell in Jupyter Notebooks in the Data Science way, it could be a good practice to establish a commonly agreed short alias (in the style of `ìmport numpy as np` or `import pandas as pd`. Something like:
`import ifcopenshell.api.run as ifc`
`ifc.root.create_entity(file, ifc_class="IfcProject")`

That altogether with some viz within the notebook from https://github.com/IfcOpenShell/ifcopenshell-jupyterhub or IFC.js could be great to have very productive experimental playgrounds.

**Comment by Moult** [comment]
I think this is now implemented. Usecases have been simplified (where possible, some are still pretty complex) to typed, standalone functions. This also allows for parsing in interpreters and IDEs (tested with pyright, pylsp, ipython, Blender, and vanilla Vim omnicompletion). No code generation is necessary now.

I generated it using just string parsing, nothing smart, so if something breaks horribly, that's probably why :)

ifcopenshell.api.run still exists and will probably hang around for a bit but I don't think there is any reason to continue using it. New users should probably go straight to using the functions directly.

![2024-05-06-145332_1357x860_scrot](https://github.com/IfcOpenShell/IfcOpenShell/assets/88302/b048d6c3-4870-46d8-8c90-49c39024f975)

![2024-05-06-145418_677x818_scrot](https://github.com/IfcOpenShell/IfcOpenShell/assets/88302/9230471c-cd7d-42a8-ac8c-c52694467cbe)

![2024-05-06-145300_902x836_scrot](https://github.com/IfcOpenShell/IfcOpenShell/assets/88302/d320c3f9-4e6a-42ce-ba3b-0085cc6319d4)
